source/common/upstream/logical_dns_cluster.cc

```diff
@@ -100,23 +100,24 @@ void LogicalDnsCluster::startResolve() {
 
   active_dns_query_ = dns_resolver_->resolve(
       dns_address, dns_lookup_family_,
-      [this, dns_address](std::list<Network::DnsResponse>&& response) -> void {
+      [this, dns_address](Network::DnsResolver::ResolutionStatus status,
+                          std::list<Network::DnsResponse>&& response) -> void {
         active_dns_query_ = nullptr;
         ENVOY_LOG(debug, "async DNS resolution complete for {}", dns_address);
-        info_->stats().update_success_.inc();
 
-        std::chrono::milliseconds refresh_rate = dns_refresh_rate_ms_;
-        if (!response.empty()) {
+        std::chrono::milliseconds final_refresh_rate = dns_refresh_rate_ms_;
+
+        // If the DNS resolver successfully resolved with an empty response list, the logical DNS
+        // cluster does not update. This ensures that a potentially previously resolved address does
+        // not stabilize back to 0 hosts.
+        if (status == Network::DnsResolver::ResolutionStatus::Success && !response.empty()) {
+          info_->stats().update_success_.inc();
           // TODO(mattklein123): Move port handling into the DNS interface.
           ASSERT(response.front().address_ != nullptr);
           Network::Address::InstanceConstSharedPtr new_address =
               Network::Utility::getAddressWithPort(*(response.front().address_),
                                                    Network::Utility::portFromTcpUrl(dns_url_));
 
-          if (respect_dns_ttl_ && response.front().ttl_ != std::chrono::seconds(0)) {
```

Comment:
moved this for consistency with the strict dns cluster code. But happy to move back up.